### PR TITLE
IMPORTANT: load VPL runtime by MSDK dispatcher

### PR DESCRIPTION
**Pull request by aobolensk**
PR description provided by @dvrogozh...

This PR highlights important change in overall Media SDK program and introduces related changes into Media SDK code base.

Media SDK lifetime comes to an end in a form it exists right now. It will be maintained to support existing applications and "legacy" platforms (see below notes on runtime). Future development is planned to happen within [oneVPL](https://github.com/oneapi-src/oneVPL) library and its runtime implementations (here is [gen runtime](https://github.com/oneapi-src/oneVPL-intel-gpu)). While the name changes, API exposed by oneVPL is actually next 2.x version of the API exposed by Media SDK. Pay attention that 2.x API is not (eventually) backward compatible with 1.x API. However, backward incompatibility comes from [1.x features which got dropped](https://spec.oneapi.com/versions/latest/elements/oneVPL/source/appendix/VPL_intel_media_sdk.html), but those structures and functions which remained has preserved their layouts and signatures – this is important to highlight for runtime discussion below.

There are few things which complicate the whole Media SDK to oneVPL migration story. Long story short, generally speaking it’s not possible to fully drop Media SDK and completely switch to oneVPL. I would like to pay attention on the following 2 aspects:

* Media SDK runtime library (libmfxhw64.so.1) is still required to support “legacy” GPU platforms since new VPL runtime (libmfx-gen.so.1) **does not** support these platforms. That’s the case, for example, for SKL/KBL/CFL.
* Some frameworks which uses Media SDK did not yet switch to oneVPL. For example, ffmpeg-qsv plugins still require Media SDK (on the contrary gstreamer msdk plugins from gst-plugins-bad set which I believe has enabled support for oneVPL already).

Overall, pay attention that oneVPL library will load different runtimes depending of the underlying GPU platform:

* It will load new VPL runtime (libmfx-gen.so.1) on a new GPU platforms (like TGL)
* It will load old VPL runtime (libmfxhw64.so.1) on an older GPU platforms (like SKL)

This PR is adding similar behavior to Media SDK dispatcher library which also will load different runtimes depending on the underlying GPU platform – see PR for details and examples.

Summarizing, the following ingredients are required to support Intel media stack thru oneVPL/MSDK:
| Repo | Produced artifacts | Comments |
| ---- | ------------------ | -------- |
| https://github.com/oneapi-src/oneVPL | libvpl.so.2, include/mfx*.h, sample_encode, sample_decode, etc. | Dispatcher library, header files and development tools. Will load libmfx-gen.so.1 on new GPUs, will load libmfxhw64.so.1 on old GPUs. Not yet supported by ffmpeg-qsv. Supported by gst-plugins-bad. |
| https://github.com/oneapi-src/oneVPL-intel-gpu | libmfx-gen.so.1 | This is new VPL runtime, supports Gen GPUs, roughly >=gen12 |
| https://github.com/Intel-Media-SDK/MediaSDK | libmfx.so.1, libmfxhw64.so.1, include/mfx*.h, sample_encode, sample_decode, etc. | Dispatcher library, **"legacy" runtime library**, header files and development tools. Will load libmfx-gen.so.1 on new GPUs, will load libmfxhw64.so.1 on old GPUs. Load **rules differ** comparing to libvpl.so.2. Supported by both ffmpeg-qsv and gst-plugins-bad. |

I hope above information will be helpful for engineers who work on Media SDK (and oneVPL) applications and for Linux distro maintainers who package or plan to package Media SDK (oneVPL). FYI: @dbermond, @jbeich, @sebastinas, @sndirsch, @tjaalton

### CODE REVIEW

File: api/mfx_dispatch/linux/mfxloader.cpp [hunk 1]
```diff
@@ -177,10 +180,19 @@ mfxStatus LoaderCtx::Init(mfxInitParam& par)
...
+  auto devices = get_devices();
+  if (devices.size()) {
+    platform = devices[0].platform;
+  }
+
   std::vector<std::string> libs;
 
-  if (MFX_IMPL_BASETYPE(par.Implementation) == MFX_IMPL_AUTO ||
-      MFX_IMPL_BASETYPE(par.Implementation) == MFX_IMPL_AUTO_ANY) {
+  if (platform >= MFX_HW_ADL_P) {
```
Comment (dvrogozh) [inline api/mfx_dispatch/linux/mfxloader.cpp:191]: We need to avoid adding new platform IDs to the list. I.e. check needs to be: if platform is unknown - load VPL.

**Review by dvrogozh — CHANGES_REQUESTED** [review]

**Comment by uartie** [comment]
What is the plan for this patch?

**Comment by uartie** [comment]
Last I heard in PXT, ADL should load VPL with MSDK dispatcher.  But last revision of this patch loads MSDK RT, afaict.

File: api/mfx_dispatch/linux/device_ids.h [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+    /* ADL-S */
+    { 0x4600, MFX_HW_ADL_S, GT_INTEGRATED },//ADL-S
+    { 0x4680, MFX_HW_ADL_S, GT_INTEGRATED },//ADL-S
+    { 0x4681, MFX_HW_ADL_S, GT_INTEGRATED },//ADL-S
+    { 0x4683, MFX_HW_ADL_S, GT_INTEGRATED },//ADL-S
+    { 0x4690, MFX_HW_ADL_S, GT_INTEGRATED },//ADL-S
+    { 0x4691, MFX_HW_ADL_S, GT_INTEGRATED },//ADL-S
+    { 0x4693, MFX_HW_ADL_S, GT_INTEGRATED },//ADL-S
+    { 0x4698, MFX_HW_ADL_S, GT_INTEGRATED },//ADL-S
+    { 0x4699, MFX_HW_ADL_S, GT_INTEGRATED },//ADL-S
```
Comment (uartie) [inline api/mfx_dispatch/linux/device_ids.h:385]: Why ADL defined?  I thought VPL was going to be loaded on ADL.  But only unknown loads VPL.

**Comment by dvrogozh** [comment]
> What is the plan for this patch?

This patch awaits for 2 things:
1. New release of https://github.com/oneapi-src/oneVPL which would fix artifacts installation paths (to follow Linux File System Standard, LFS)
2. First release of new runtime, i.e. https://github.com/oneapi-src/oneVPL-intel-gpu

W/o 2nd prerequisites we risk to make a release which would introduce dependency from a new runtime which actually don't have any release yet - which would not be a good thing. And 1st prerequisite is just needed to tell complete story to community (and Linux distro maintainers at the first place) where we go with msdk and vpl.

In summary, we are trying to level set few things before actually introducing important changes and approaching community.

File: api/mfx_dispatch/linux/device_ids.h (hunk 2 above)
Comment (dvrogozh) [reply api/mfx_dispatch/linux/device_ids.h:385]: There are 2 platforms for ADL: ADL-S and ADL-P. Mediasdk runtime ends up with ADL-S (inclusively), but it does not support ADL-P. FYI: VPL runtime supports platforms starting from TGL (aka gen 12 family). The story here is somewhat complicated, we plan to provide some documentation around that. Hold on.

File: api/mfx_dispatch/linux/device_ids.h [hunk 3]
```diff
@@ -0,0 +1,471 @@
+#include <stdio.h>
```
Comment (dvrogozh) [inline api/mfx_dispatch/linux/device_ids.h:1]: License is missed!

File: api/mfx_dispatch/linux/mfxloader.cpp [hunk 4]
```diff
@@ -52,6 +54,7 @@ namespace MFX {
     #else
         #define LIBMFXSW "libmfxsw64.so.1"
         #define LIBMFXHW "libmfxhw64.so.1"
+        #define ONEVPLRT "libmfx-gen.so.1.2"
```
Comment (dvrogozh) [inline api/mfx_dispatch/linux/mfxloader.cpp:57]: Remove `.2`, we need to load library via soname.

File: tests/unit/suites/mfx_dispatch/linux/mfx_dispatch_test_cases_libs.cpp [hunk 5]
```diff
@@ -84,15 +85,24 @@ TEST_P(DispatcherLibsTestParametrized, ShouldEnumerateCorrectLibNames)
...
-    if (MFX_IMPL_BASETYPE(impl) == MFX_IMPL_AUTO ||
-        MFX_IMPL_BASETYPE(impl) == MFX_IMPL_AUTO_ANY)
-    {
+    eMFXHWType platform = MFX_HW_UNKNOWN;
+    auto devices = get_devices();
+    if (devices.size()) {
+        platform = devices[0].platform;
+    }
+
+    if (platform == MFX_HW_UNKNOWN) {
```
Comment (dvrogozh) [inline tests/unit/suites/mfx_dispatch/linux/mfx_dispatch_test_cases_libs.cpp:101]: Add a code comment that we load VPL for all unknown platforms which correspond to future platforms. And make condition upside down, i.e.:
```
if (platform != MFX_HW_UNKNOWN) {
  load.emplace_back(l"ibmfxhw64.so");
} else {
  load.emplace_back(l"ibvpl64.so");
}
```

File: api/mfx_dispatch/linux/device_ids.h [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+};
+
+typedef struct {
+    int          device_id;
+    eMFXHWType   platform;
+    GraphicsType type;
+} mfx_device_item;
+
+// list of legal dev ID for Intel's graphics
+const mfx_device_item listLegalDevIDs[] = {
```
Comment (dvrogozh) [inline api/mfx_dispatch/linux/device_ids.h:53]: s/listLegalDevIDs/msdkDevIDs/

File: api/mfx_dispatch/linux/device_ids.h [hunk 7]
```diff
@@ -0,0 +1,471 @@
+#include <stdio.h>
+#include <sys/stat.h>
```
Comment (dvrogozh) [inline api/mfx_dispatch/linux/device_ids.h:2]: add <vector>

**Comment by dvrogozh** [comment]
VPL runtime got the first release: https://github.com/oneapi-src/oneVPL-intel-gpu/releases/tag/intel-onevpl-21.2.3. @aobolensk : please, address comments and promote to official PR.

File: api/mfx_dispatch/linux/device_ids.h (hunk 3 above)
Comment (aobolensk) [reply api/mfx_dispatch/linux/device_ids.h:21]: Added license

File: api/mfx_dispatch/linux/mfxloader.cpp (hunk 4 above)
Comment (aobolensk) [reply api/mfx_dispatch/linux/mfxloader.cpp]: Removed

File: tests/unit/suites/mfx_dispatch/linux/mfx_dispatch_test_cases_libs.cpp [hunk 8]
```diff
@@ -84,15 +85,24 @@ TEST_P(DispatcherLibsTestParametrized, ShouldEnumerateCorrectLibNames)
 #elif defined(__x86_64__)
     const std::string LIBMFXSW("libmfxsw64.so.1");
     const std::string LIBMFXHW("libmfxhw64.so.1");
+    const std::string ONEVPLRT("libmfx-gen.so.1.2");
```
[the remark this answers is not in the document]
Comment (aobolensk) [reply tests/unit/suites/mfx_dispatch/linux/mfx_dispatch_test_cases_libs.cpp]: Removed

File: tests/unit/suites/mfx_dispatch/linux/mfx_dispatch_test_cases_libs.cpp (hunk 5 above)
Comment (aobolensk) [reply tests/unit/suites/mfx_dispatch/linux/mfx_dispatch_test_cases_libs.cpp]: Done

File: api/mfx_dispatch/linux/device_ids.h (hunk 6 above)
Comment (aobolensk) [reply api/mfx_dispatch/linux/device_ids.h]: Done

File: api/mfx_dispatch/linux/device_ids.h (hunk 7 above)
Comment (aobolensk) [reply api/mfx_dispatch/linux/device_ids.h:22]: Done

File: api/mfx_dispatch/linux/device_ids.h [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+        if (!file) break;
+        if (fscanf(file, "%x", &device.device_id) == EOF) break;
+        fclose(file);
+        device.platform = get_platform(device.device_id);
+        device.type = get_graphics_type(device.device_id);
+        result.emplace_back(device);
+    }
+    std::sort(result.begin(), result.end(), [](const Device &a, const Device &b) {
+        if (a.type == b.type)
+            return a.device_id < b.device_id;
```
Comment (dvrogozh) [inline api/mfx_dispatch/linux/device_ids.h:490]: Need to fix sorting.... It's not guaranteed, I think, that new devices will have devids > than existing ones. For example, see APL=0x0a84 vs. HSW=0x0D1A.

For what we care here in dispatcher, we can probably just sort per `eMFXHWType platform`. The only thing we actually need from such sort is: do we have new platforms on the system or not?

Comment (aobolensk) [reply api/mfx_dispatch/linux/device_ids.h]: Done sorting per platform

File: README.md [hunk 10]
```diff
@@ -9,6 +9,47 @@ Media SDK is a part of Intel software stack for graphics:
...
+(libmfx-gen.so.1) comes with the support of new Gen platforms.
+
+Pay attention that Intel® Media SDK has forward compatibility with new VPL
+runtime (libmfx-gen.so.1) in the scope of API features supported by **both** 1.x
+and 2.x API series. As such, if application is built against Intel® Media
+SDK, it still can work on new platforms. For that purpose Media SDK Dispatcher
+loads either MSDK Legacy Runtime (libmfxhw64.so.1) or VPL Runtime (libmfx-gen.so.1)
+depending on the underlying platform. See support matrix below.
+
+| Platform                               | libmfxhw64.so.1 | libmfx-gen.so.1 |
```
Comment (dvrogozh) [inline README.md:35]: I think these column names might be confusing. People might think that this is the table of what particular runtime actually support, but it is actually not. Instead, this is a table of what libmfx.so.1 will load depending on the platform. Need to fix.

File: api/mfx_dispatch/linux/device_ids.h (hunk 2 above)
Comment (uartie) [reply api/mfx_dispatch/linux/device_ids.h:408]: Does ADL-S and ADL-P have the same GPU Gen?

Comment (dmitryermilov) [reply api/mfx_dispatch/linux/device_ids.h:408]: Yes, It's Gen12

Comment (uartie) [reply api/mfx_dispatch/linux/device_ids.h:408]: Sorry, what I meant was, is it same minor rev, too (12.2)?

Comment (dmitryermilov) [reply api/mfx_dispatch/linux/device_ids.h:408]: Yes

File: README.md [hunk 11]
```diff
@@ -9,6 +9,47 @@ Media SDK is a part of Intel software stack for graphics:
...
+| BXT (Broxton) / APL (Apollo Lake)      | ✔               |                 |
+| KBLx                                   | ✔               |                 |
+| ICL (Ice Lake)                         | ✔               |                 |
+| JSL (Jasper Lake) / EHL (Elkhart Lake) | ✔               |                 |
+| TGL (Tiger Lake)                       | ✔               |                 |
+| DG1 (Xe MAX)/SG1                       | ✔               |                 |
+| RKL (Rocket Lake)                      | ✔               |                 |
+| ADL-S (Alder Lake S)                   | ✔               |                 |
+| ADL-P (Alder Lake P)                   |                 | ✔               |
+| Future platforms...                    |                 | ✔               |
```
Comment (dmitryermilov) [inline README.md:48]: Need to consider the case when on system which supports both MSDK and VPL (e.g. TGL), only VPL RT is installed by a user.

**Review by dmitryermilov — COMMENTED** [review]
Also, I assume to give a complete message we need to cover Windows path. What do you think?

File: README.rst [hunk 12]
```diff
@@ -0,0 +1,298 @@
+Intel® Media SDK
```
Comment (dvrogozh) [inline README.rst:1]: Can you, please, cherry-pick 2 commits and not squash them into one from here: https://github.com/dvrogozh/MediaSDK/commits/docs2. I really believe that things needs to be done step by step. First change .md to .rst, then - add new info. Otherwise new info is obscured.

**Review by dvrogozh — CHANGES_REQUESTED** [review]

File: api/mfx_dispatch/linux/device_ids.h [hunk 13]
```diff
@@ -0,0 +1,494 @@
...
+        file = fopen(path.c_str(), "r");
+        if (!file) break;
+        if (fscanf(file, "%x", &device.device_id) == EOF) break;
+        fclose(file);
+        device.platform = get_platform(device.device_id);
+        device.type = get_graphics_type(device.device_id);
+        result.emplace_back(device);
+    }
+    std::sort(result.begin(), result.end(), [](const Device &a, const Device &b) {
+        if (a.type == b.type)
```
[the remark this answers is not in the document]
Comment (aobolensk) [reply api/mfx_dispatch/linux/device_ids.h]: No, we do not use it. Removed

File: api/mfx_dispatch/linux/device_ids.h [hunk 14]
```diff
@@ -0,0 +1,494 @@
...
+            return a.platform < b.platform;
+        return a.type < b.type;
```
[the remark this answers is not in the document]
Comment (aobolensk) [reply api/mfx_dispatch/linux/device_ids.h]: Dropped type comparison

File: README.rst (hunk 12 above)
Comment (aobolensk) [reply README.rst:1]: Done

File: README.rst [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+| RKL (Rocket Lake)    | Legacy       | ✔               | ✔               | ✔               |                 |
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+| ADL-S (Alder Lake S) | Legacy       | ✔               | ✔               | ✔               |                 |
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+| ADL-P (Alder Lake P) | VPL          |                 | ✔               |                 | ✔               |
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+| Future platforms...  | VPL          |                 | ✔               |                 | ✔               |
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+| Multi GPU            | Legacy + VPL | See above                         | ✔               | via env var [2] |
```
Comment (aobolensk) [inline README.rst:79]: Dispatcher loads VPL in multi GPU case and using MSDK_RUNTIME environment variable is not implemented yet.

File: README.rst [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+
+Pay attention that Intel® Media SDK lifetime comes to an end in a form it
+exists right now. In particular, 
+
+* API 1.35 is projected to be the last API of 1.x API series
+* Runtime library (libmfxhw64.so.1) is not planned to get support of new Gen platforms
+* Project is going to be supported in maintainence mode, critical fixes only
+
+All future development is planned to happen within
+[oneVPL](https://github.com/oneapi-src/oneVPL) library and its runtime
```
Comment (dvrogozh) [inline README.rst:31]: fix link, that's .md syntax, need .rst syntax as follows:
```
`oneVPL <https://github.com/oneapi-src/oneVPL>`_
```

File: README.rst [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+implementations which are direct successors of Intel® Media SDK. oneVPL introduces
+API 2.x series which is not backward compatible with API 1.x series (some
+features got dropped). New [VPL Runtime for Gen Graphics](https://github.com/oneapi-src/oneVPL-intel-gpu)
```
Comment (dvrogozh) [inline README.rst:34]: same:
```
`VPL Runtime for Gen Graphics <https://github.com/oneapi-src/oneVPL-intel-gpu>`_
```

File: api/mfx_dispatch/linux/mfxloader.cpp [hunk 18]
```diff
@@ -177,23 +180,35 @@ mfxStatus LoaderCtx::Init(mfxInitParam& par)
     return MFX_ERR_UNSUPPORTED;
   }
 
+  eMFXHWType platform = MFX_HW_UNKNOWN;
+  auto devices = get_devices();
+  if (devices.size()) {
+    platform = devices[0].platform;
```
Comment (dvrogozh) [inline api/mfx_dispatch/linux/mfxloader.cpp:186]: You sorted deive list per platform, so MFX_HW_UNKNOWN will be the [0] one. And you will load VPL while we need to load MSDK runtime. Please, fix.

File: README.rst [hunk 19]
```diff
@@ -0,0 +1,301 @@
...
+(libmfx-gen.so.1) comes with the support of new Gen platforms.
+
+Pay attention that Intel® Media SDK has forward compatibility with new VPL
+runtime (libmfx-gen.so.1) in the scope of API features supported by **both** 1.x
+and 2.x API series. As such, if application is built against Intel® Media
+SDK, it still can work on new platforms. For that purpose Media SDK Dispatcher
+loads either MSDK Legacy Runtime (libmfxhw64.so.1) or VPL Runtime (libmfx-gen.so.1)
```
Comment (dmitryermilov) [inline README.rst:41]: MSDK Legacy Runtime- > SDK Legacy Runtime or Media SDK Legacy Runtime

File: README.rst [hunk 20]
```diff
@@ -0,0 +1,301 @@
...
+exists right now. In particular, 
+
+* API 1.35 is projected to be the last API of 1.x API series
+* Runtime library (libmfxhw64.so.1) is not planned to get support of new Gen platforms
+* Project is going to be supported in maintainence mode, critical fixes only
+
+All future development is planned to happen within
+[oneVPL](https://github.com/oneapi-src/oneVPL) library and its runtime
+implementations which are direct successors of Intel® Media SDK. oneVPL introduces
+API 2.x series which is not backward compatible with API 1.x series (some
```
Comment (dmitryermilov) [inline README.rst:33]: Shall we add a link to https://software.intel.com/content/www/us/en/develop/articles/upgrading-from-msdk-to-onevpl.html . It has a list of these features.

File: api/mfx_dispatch/linux/device_ids.h [hunk 21]
```diff
@@ -0,0 +1,475 @@
...
+    MFX_HW_DG1       = 0x1600003,
+    MFX_HW_ADL_S     = MFX_HW_TGL_LP + 4,
+};
+
+typedef struct {
+    int          device_id;
+    eMFXHWType   platform;
+} mfx_device_item;
+
+// list of legal dev ID for Intel's graphics
```
Comment (dmitryermilov) [inline api/mfx_dispatch/linux/device_ids.h:69]: I believe better to remove this comment: "// list of legal dev ID for Intel's graphics". It gives an impression, there are no other legal Intel graphics platfroms

File: api/mfx_dispatch/linux/device_ids.h [hunk 22]
```diff
@@ -0,0 +1,475 @@
...
+    case MFX_HW_JSL: return "JSL";
+    case MFX_HW_EHL: return "EHL";
+    case MFX_HW_TGL_LP: return "TGL_LP";
+    case MFX_HW_DG1: return "DG1";
+    case MFX_HW_RKL: return "RKL";
+    case MFX_HW_ADL_S: return "ADL_S";
+    case MFX_HW_UNKNOWN: default: return "UNKNOWN";
+    }
+}
+
```
Comment (dmitryermilov) [inline api/mfx_dispatch/linux/device_ids.h:431]: Is it used?

File: api/mfx_dispatch/linux/device_ids.h [hunk 23]
```diff
@@ -0,0 +1,475 @@
...
+        if (!file) break;
+        if (fscanf(file, "%x", &device.vendor_id) == EOF) break;
+        fclose(file);
+        if (device.vendor_id != 0x8086) {  // Filter out non-Intel devices
+            continue;
+        }
+        path = std::string(dir) + "/renderD" + std::to_string(128 + i) + device_id_file;
+        file = fopen(path.c_str(), "r");
+        if (!file) break;
+        if (fscanf(file, "%x", &device.device_id) == EOF) break;
```
Comment (dmitryermilov) [inline api/mfx_dispatch/linux/device_ids.h:466]: shall we also call close()  here?

File: README.rst [hunk 24]
```diff
@@ -0,0 +1,301 @@
...
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+| Future platforms...  | VPL          |                 | ✔               |                 | ✔               |
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+| Multi GPU            | Legacy + VPL | See above                         | ✔               | via env var [2] |
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+
+**Notes**:
+
+* [1] KBLx is one of: KBL (Kaby Lake), CFL (Coffe Lake), WHL (Whiskey Lake), CML (Comet Lake), AML (Amber Lake)
+* [2] On the multi GPU system which has both "Legacy" and "VPL" GPUs Media SDK Dispatcher loads MSDK Legacy
```
Comment (dmitryermilov) [inline README.rst:85]: MSDK Legacy -> "SDK Legacy" or "Media SDK Legacy"

File: README.rst [hunk 25]
```diff
@@ -0,0 +1,301 @@
...
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+| DG1 (Xe MAX)         | Legacy       | ✔               | ✔               | ✔               |                 |
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+| SG1                  | Legacy       | ✔               |                 | ✔               |                 |
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+| RKL (Rocket Lake)    | Legacy       | ✔               | ✔               | ✔               |                 |
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+| ADL-S (Alder Lake S) | Legacy       | ✔               | ✔               | ✔               |                 |
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+| ADL-P (Alder Lake P) | VPL          |                 | ✔               |                 | ✔               |
```
Comment (dmitryermilov) [inline README.rst:75]: It needs separate discussion. @daleksan , please help to find a conclusion next week.

**Review by dmitryermilov — CHANGES_REQUESTED** [review]
Intentionally putting -1 before we clarify ADL

File: README.rst (hunk 20 above)
Comment (dvrogozh) [reply README.rst:33]: Is there a link to VPL spec instead? I guess it might have such information as well. I would prefer to refer to a more formal spec here, but we can also add a list of useful links for more reading in the end of whole "Matrix" paragraph.

Comment (dmitryermilov) [reply README.rst:33]: https://spec.oneapi.com/versions/latest/elements/oneVPL/source/index.html

File: README.rst (hunk 16 above)
Comment (aobolensk) [reply README.rst]: Fixed

File: README.rst (hunk 17 above)
Comment (aobolensk) [reply README.rst]: Fixed

File: api/mfx_dispatch/linux/mfxloader.cpp (hunk 18 above)
Comment (aobolensk) [reply api/mfx_dispatch/linux/mfxloader.cpp]: Fixed. Now devices[devices.size() - 1] is used to automatically select legacy path if MSDK and VPL platform are detected at the same time

File: README.rst (hunk 19 above)
Comment (aobolensk) [reply README.rst]: Fixed

File: api/mfx_dispatch/linux/device_ids.h (hunk 21 above)
Comment (aobolensk) [reply api/mfx_dispatch/linux/device_ids.h]: Fixed

File: api/mfx_dispatch/linux/device_ids.h (hunk 22 above)
Comment (aobolensk) [reply api/mfx_dispatch/linux/device_ids.h]: No. Removed

File: api/mfx_dispatch/linux/device_ids.h (hunk 23 above)
Comment (aobolensk) [reply api/mfx_dispatch/linux/device_ids.h]: Yes. Fixed it

File: README.rst (hunk 24 above)
Comment (aobolensk) [reply README.rst]: Fixed

File: api/mfx_dispatch/linux/mfxloader.cpp [hunk 26]
```diff
@@ -177,23 +180,38 @@ mfxStatus LoaderCtx::Init(mfxInitParam& par)
...
-    libs.emplace_back(LIBMFXSW);
-    libs.emplace_back(MFX_MODULES_DIR "/" LIBMFXSW);
-  } else if (par.Implementation & MFX_IMPL_HARDWARE ||
-             par.Implementation & MFX_IMPL_HARDWARE_ANY) {
-    libs.emplace_back(LIBMFXHW);
-    libs.emplace_back(MFX_MODULES_DIR "/" LIBMFXHW);
-  } else if (par.Implementation & MFX_IMPL_SOFTWARE) {
-    libs.emplace_back(LIBMFXSW);
-    libs.emplace_back(MFX_MODULES_DIR "/" LIBMFXSW);
+  const char *selected_runtime = getenv("MSDK_RUNTIME");
```
Comment (dmitryermilov) [inline api/mfx_dispatch/linux/mfxloader.cpp:191]: https://github.com/Intel-Media-SDK/MediaSDK/pull/2728/files: INTEL_MEDIA_RUNTIME

Comment (aobolensk) [reply api/mfx_dispatch/linux/mfxloader.cpp]: Renamed

File: README.rst [hunk 27]
```diff
@@ -0,0 +1,301 @@
...
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+
+**Notes**:
+
+* [1] KBLx is one of: KBL (Kaby Lake), CFL (Coffe Lake), WHL (Whiskey Lake), CML (Comet Lake), AML (Amber Lake)
+* [2] On the multi GPU system which has both "Legacy" and "VPL" GPUs Media SDK Dispatcher loads Media SDK Legacy
+  Runtime (libmfxhw64.so.1) by default. VPL Runtime (libmfx-gen.so.1) can be loaded if the following environment
+  variable will be specified::
+
+    export MSDK_RUNTIME=libmfx-gen.so.1
```
Comment (dvrogozh) [inline README.rst:89]: See #2728: we need to rename env. variable to `INTEL_MEDIA_RUNTIME` which should have 2 possible values: `MSDK|ONEVPL`.

Comment (aobolensk) [reply README.rst]: Done

File: README.rst [hunk 28]
```diff
@@ -0,0 +1,303 @@
...
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+| TGL (Tiger Lake)     | Legacy       | ✔               | ✔               | ✔               |                 |
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+| DG1 (Xe MAX)         | Legacy       | ✔               | ✔               | ✔               |                 |
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+| SG1                  | Legacy       | ✔               |                 | ✔               |                 |
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+| RKL (Rocket Lake)    | Legacy       | ✔               | ✔               | ✔               |                 |
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+| ADL-S (Alder Lake S) | Legacy       | ✔               | ✔               | ✔               |                 |
```
Comment (daleksan) [inline README.rst:73]: We need to remove ADL support on MediaSDK because it will be covered by VPL

**Review by daleksan — CHANGES_REQUESTED** [review]

File: README.rst (hunk 20 above)
Comment (dvrogozh) [reply README.rst:33]: I think here it is:
* https://spec.oneapi.com/versions/latest/elements/oneVPL/source/appendix/VPL_intel_media_sdk.html

Comment (dmitryermilov) [reply README.rst:33]: It's a good link

File: api/mfx_dispatch/linux/device_ids.h (hunk 2 above)
Comment (dmitryermilov) [reply api/mfx_dispatch/linux/device_ids.h]: @uartie , please see reply from @daleksan  below.

File: README.rst (hunk 28 above)
Comment (dmitryermilov) [reply README.rst:73]: @aobolensk , please fix

Comment (aobolensk) [reply README.rst]: Done. Removed support for ADL-S and fixed README

**Review by dmitryermilov — APPROVED** [review]

File: README.rst (hunk 20 above)
Comment (dvrogozh) [reply README.rst:33]: @aobolensk : can you, please, add the link here:
```
some `features got dropped <https://spec.oneapi.com/versions/latest/elements/oneVPL/source/appendix/VPL_intel_media_sdk.html>`_
```

Comment (aobolensk) [reply README.rst:33]: Done

**Review by dmitryermilov — APPROVED** [review]

File: README.rst [hunk 29]
```diff
@@ -0,0 +1,304 @@
...
+| Future platforms...  | VPL          |                 | ✔               |                 | ✔               |
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+| Multi GPU            | Legacy + VPL | See above                         | ✔               | via env var [2] |
++----------------------+--------------+-----------------+-----------------+-----------------+-----------------+
+
+**Notes**:
+
+* [1] KBLx is one of: KBL (Kaby Lake), CFL (Coffe Lake), WHL (Whiskey Lake), CML (Comet Lake), AML (Amber Lake)
+* [2] On the multi GPU system which has both "Legacy" and "VPL" GPUs Media SDK Dispatcher loads Media SDK Legacy
+  Runtime (libmfxhw64.so.1) by default. VPL Runtime (libmfx-gen.so.1) can be loaded if the following environment
```
Comment (dvrogozh) [inline README.rst:87]: Rewrite:
```
VPL Runtime (libmfx-gen.so.1) or Media SDK Runtime (libmfxhw64.so.1) can be explicitly selected via the following environment variable::

  export INTEL_MEDIA_RUNTIME=ONEVPL  # for VPL Runtime: libmfx-gen.so.1
  or
  export INTEL_MEDIA_RUNTIME=MSDK # for Media SDK Runtime: libmfxhw64.so.1

```

Comment (aobolensk) [reply README.rst]: Done

**Review by dvrogozh — APPROVED** [review]

**Review by dmitryermilov — APPROVED** [review]

**Review by daleksan — APPROVED** [review]

**Comment by dvrogozh** [comment]
> What is the plan for this patch?

@uartie : fyi, we plan to merge this PR early next week.
